zkvm/src/vm.rs

```diff
@@ -586,90 +600,28 @@ where
     }
 
     fn spend_input(&mut self, input: Input) -> Result<(Contract, UTXO), VMError> {
-        match input {
-            Input::Opaque(data) => self.decode_input(data),
-            Input::Witness(_) => unimplemented!(),
-        }
-    }
-
-    fn decode_input(&mut self, data: Vec<u8>) -> Result<(Contract, UTXO), VMError> {
-        // Input  =  PreviousTxID || PreviousOutput
-        // PreviousTxID  =  <32 bytes>
-        let mut slice = Subslice::new(&data);
-        let txid = TxID(slice.read_u8x32()?);
-        let output_slice = &slice;
-        let contract = self.decode_output(slice)?;
-        let utxo = UTXO::from_output(output_slice, &txid);
-        Ok((contract, utxo))
-    }
-
-    fn decode_output<'a>(&mut self, mut output: Subslice<'a>) -> Result<Contract, VMError> {
-        //    Output  =  Predicate  ||  LE32(k)  ||  Item[0]  || ... ||  Item[k-1]
-        // Predicate  =  <32 bytes>
-        //      Item  =  enum { Data, Value }
-        //      Data  =  0x00  ||  LE32(len)  ||  <bytes>
-        //     Value  =  0x01  ||  <32 bytes> ||  <32 bytes>
-
-        let predicate = Predicate::opaque(output.read_point()?);
-        let k = output.read_size()?;
-
-        // sanity check: avoid allocating unreasonably more memory
-        // just because an untrusted length prefix says so.
-        if k > output.len() {
-            return Err(VMError::FormatError);
-        }
-
-        let mut payload: Vec<PortableItem> = Vec::with_capacity(k);
-        for _ in 0..k {
-            let item = match output.read_u8()? {
-                DATA_TYPE => {
-                    let len = output.read_size()?;
-                    let bytes = output.read_bytes(len)?;
-                    PortableItem::Data(Data::Opaque(bytes.to_vec()))
+        let contract = self.unfreeze_contract(input.contract);
+        Ok((contract, input.utxo))
+    }
+
+    fn unfreeze_contract(&mut self, contract: FrozenContract) -> Contract {
+        let payload = contract
+            .payload
+            .into_iter()
+            .map(|p| match p {
+                FrozenItem::Data(d) => PortableItem::Data(d),
+                FrozenItem::Value(v) => {
+                    let qty = self.make_variable(v.qty);
+                    let flv = self.make_variable(v.flv);
+                    let val = Value { qty, flv };
+                    PortableItem::Value(val)
```

Comment:
alternative this can be written with less intermediate variables for clarity:

```rust
PortableItem::Value(Value {
    qty: self.make_variable(v.qty),
    flv: self.make_variable(flv),
})
```